kernel/src/driver_manager.rs

```diff
@@ -0,0 +1,126 @@
+use alloc::{boxed::Box, collections::LinkedList, sync::Arc};
+
+use super::device_tree::DeviceTree;
+use super::globals;
+use super::mm;
+use super::paging::PagingImpl as _;
+use drivers::{Console, Driver};
+
+pub struct DriverManager {
+    drivers: LinkedList<Arc<dyn Driver>>,
+}
+
+impl DriverManager {
+    fn new() -> Self {
+        Self {
+            drivers: LinkedList::new(),
+        }
+    }
+
+    pub fn with_devices(dt: &DeviceTree) -> Self {
+        let mut mgr = Self::new();
+
+        let cons_node = dt.console_node();
+        if let Some(cons_node) = &cons_node {
+            map_dt_regions(cons_node);
+            let cons_driver = matching_console_driver(cons_node);
+            if let Some(cons_driver) = cons_driver {
+                mgr.register_console(cons_driver);
+            } else {
+                unmap_dt_regions(cons_node);
+            }
+        }
+
+        mgr
+    }
+
+    fn register_console(&mut self, cons: Box<dyn Console + Sync + Send>) {
+        let cons: Arc<dyn Console + Sync + Send> = Arc::from(cons);
+        self.register_driver(cons.clone());
+        globals::CONSOLE.set(cons.clone()).unwrap();
+    }
+
+    fn register_driver(&mut self, drv: Arc<dyn Driver>) {
+        self.drivers.push_back(drv);
+    }
+}
+
+pub fn matching_console_driver(
+    node: &fdt::node::FdtNode,
+) -> Option<Box<dyn Console + Sync + Send>> {
+    let compatible = node
+        .properties()
+        .find(|prop| prop.name == "compatible")
+        .and_then(|some_prop| some_prop.as_str())
+        .unwrap_or("");
+
+    let compatibles = compatible.split('\0');
+
+    let console_base = node.reg()?.next()?.starting_address as usize;
+
+    for compat in compatibles {
+        match compat {
+            "arm,pl011" => return Some(drivers::pl011::Pl011::boxed(console_base)),
+            // "ns16550" => return Some(&drivers::ns16550::Ns16550::new(console_base)),
+            _ => continue,
+        }
+    }
+
+    None
+}
+
+fn map_dt_regions(node: &fdt::node::FdtNode) {
+    let pagesize = crate::PagingImpl::get_page_size();
+
+    if let Some(reg) = node.reg() {
+        for memory_region in reg {
+            let start = memory_region.starting_address as usize;
+            let size = memory_region.size.unwrap_or(pagesize);
```

Comment:
If there is no size, it should be an error, right?

Reply:
Good point, but now I'm just confused about the parser, shouldn't it error out earlier if there is no size ?

Reply:
I think we should just panic instead of using a default value

Reply:
I think we should just panic instead of using a default value